### PR TITLE
feat(protocol-engine): Implement Magnetic Module engage

### PR DESCRIPTION
**Pull request by SyntaxColoring**
# Overview

This implements a Protocol Engine command to engage a Magnetic Module's magnets. This closes part of #8243.

# Changelog

* Hardware API: Publicly expose the method that `MagDeck` has been using to validate an engage height. This lets Protocol Engine validate engage heights in analysis without instantiating an entire simulated Magnetic Module.
* ModuleView`: Add a new method, `calculate_magnet_hardware_height()`, to convert a human-friendly height in true mm to a Magnetic-Module–friendly height in model-specific units. We already had the reverse, sort of, in `calculate_magnet_height()`.
* `ModuleView`: Add a new method, `find_loaded_hardware_module()`, to seek a matching hardware module object in the list of attached modules returned by the hardware controller, given its Protocol Engine module ID. Rename the older `find_attached_module()` to `select_hardware_module_to_load()` to avoid confusion.
* Protocol Engine commands: Give command implementations direct access to the Protocol Engine's `StateView` and `HardwareControlAPI`. This means implementing a command no longer needs to go through one of the middleman classes like `EquipmentHandler`, when that indirection is unnecessary.
* Implement the `magneticModule/engageMagnet` command, whose skeleton was laid out in a prior PR.

# Review requests

* Naming suggestions?
    * **Known problem:** This leaves two distinct notions of "hardware module" in `ModuleView`: the `HardwareModule` class, which is not actually a hardware module object from the hardware controller, and `opentrons.hardware_control.modules.AbstractModule`, which is.
* Error-handling test suggestions (error propagation)?
* See manual test plan, below.

# Manual test plan

We need to manually test:

1. Engage commands get routed to the correct physical module, when there are multiple.
2. Engage commands move the magnets to the correct physical height.
3. Errors propagate as expected.

To do this:

0. Connect a couple of Temperature Modules to the robot. Make them different generations if you can, to make things interesting.

0. Create a basic run:

   ```
   POST /runs
   
   {
       "data": {}
   }
   ```

0. Load both modules:

   ```
   POST /runs/{{run_id}}/commands?waitUntilComplete=true
   
   {
       "data": {
           "commandType:" "loadModule",
           "params": {
               "moduleId": "mm1",
               "model": "magneticModuleV1",
               "location": {
                   "slotName": "1" 
               }
           }
       }
   }
   
   [Repeat for the second module. Remember to change params as appropriate.]
   ```

0. Issue engage commands and confirm that the magnets reach the commanded height.

   ```
   POST /runs/{{run_id}}/commands?waitUntilComplete=true
   
   {
       "data": {
           "commandType": "magneticModule/engageMagnet",
           "params": {
               "moduleId": "mm1",
               "engageHeight": 10
           }
           
       }
   }
   ```
   
   Given historical bugs related to units and origin point, I strongly recommend measuring the actual physical height with calipers. I found that a GEN1 module was moving ~ 1 mm too low, which I think is because of https://github.com/Opentrons/opentrons/issues/9585.

0. Conform that the command fails if you provide a bogus `moduleId`.
0. Confirm that the command fails if you provide an out-of-range `engageHeight`.
0. Confirm that the engage command fails if you physically unplug the module before running the engage command, but after successfully running the load command.
0. If you have another type of module (non-Magnetic) handy, load it and confirm that the engage command fails if you provide it that module's ID).

# Risk assessment

Low: this is new code, so it shouldn't break anything in production.

### CODE REVIEW

File: api/src/opentrons/protocol_engine/execution/equipment.py [hunk 1]
```diff
@@ -227,3 +233,28 @@ async def load_module(
...
             definition=attached_module.definition,
         )
+
+    async def engage_magnets(
+        self,
+        magnetic_module_id: str,
+        mm_above_labware_base: float,
+    ) -> None:
+        """Engage a loaded Magnetic Module's magnets."""
+        # Will raise if the given module ID hasn't been loaded.
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/equipment.py]: Might be worth putting these in the docstring rather than inline comments

File: api/src/opentrons/protocol_engine/execution/equipment.py [hunk 2]
```diff
@@ -227,3 +233,28 @@ async def load_module(
             serial_number=attached_module.serial_number,
             definition=attached_module.definition,
         )
+
+    async def engage_magnets(
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/equipment.py]: Method looks good, but I don't think the `EquipmentHandler` (described as `Implementation logic for labware, pipette, and module loading.`) is the best home for it, given that this class is sort of too long already IMO. I would either:

- Create a new `MagneticModuleHandler` (or something along those lines)
    - Same path we took in #9569 
- Start moving towards the "less indirection" world we've been talking about
    1. Give `EngageImplementation` access to a `HardwareAPI` and a `StateView`
    2. Move the contents of this method as written into `EngageImplementation.execute`

File: api/src/opentrons/protocol_engine/execution/equipment.py [hunk 3]
```diff
@@ -227,3 +233,28 @@ async def load_module(
...
+        # Will raise if the given module ID hasn't been loaded.
+        model = self._state_store.modules.get_model(module_id=magnetic_module_id)
+
+        # Will raise if either:
+        #   * The given module ID points to a module that's not a Magnetic Module.
+        #   * The given height is invalid.
+        hardware_height = self._state_store.modules.calculate_magnet_hardware_height(
+            magnetic_module_model=model,
+            mm_above_labware_base=mm_above_labware_base,
+        )
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/equipment.py]: We've got two calls in a row to the same state view; I think they could be combined into one call

File: api/src/opentrons/protocol_engine/execution/equipment.py [hunk 4]
```diff
@@ -227,3 +233,28 @@ async def load_module(
             serial_number=attached_module.serial_number,
             definition=attached_module.definition,
         )
+
+    async def engage_magnets(
+        self,
+        magnetic_module_id: str,
+        mm_above_labware_base: float,
+    ) -> None:
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/equipment.py]: If we wanted to, I think returning the actually engaged height would be useful. Not super important, but it's nice to return data where there is potentially useful data to return

File: api/tests/opentrons/protocol_engine/commands/magnetic_module/test_engage.py [hunk 5]
```diff
@@ -1,37 +1,32 @@
...
     )
-    await subject.execute(params=params)
+    result = await subject.execute(params=params)
+    decoy.verify(
+        await equipment.engage_magnets(
+            magnetic_module_id="module-id",
+            mm_above_labware_base=3.14159,
+        )
+    )
+    assert result == EngageResult()
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/commands/magnetic_module/test_engage.py]: If we returned data, this could look like:

```suggestion
    decoy.when(
        await equipment.engage_magnets(
            magnetic_module_id="module-id",
            mm_above_labware_base=3.14159,
        )
    ).then_return(42)

    result = await subject.execute(params=params)

    assert result == EngageResult(mmFromBase=42)
```

File: api/tests/opentrons/protocol_engine/commands/magnetic_module/test_engage.py [hunk 6]
```diff
@@ -1,37 +1,32 @@
...
     params = EngageParams(
         moduleId="module-id",
         engageHeight=3.14159,
     )
-    await subject.execute(params=params)
+    result = await subject.execute(params=params)
+    decoy.verify(
+        await equipment.engage_magnets(
+            magnetic_module_id="module-id",
+            mm_above_labware_base=3.14159,
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/commands/magnetic_module/test_engage.py]: A couple bits of confusion here we noticed:

- `above` might be confusing because you can engage below the base height
    - Suggestion: maybe `from`, given "more positive" already means "up" in our existing coordinate system 
- `labware_base` might be confusing because not all labware have bases (e.g. non/half-skirted labware)
    - I guess we're sort of talking about the magdeck's "floor", for lack of better terminology
    - Suggestion: Maybe just `base`?

File: api/src/opentrons/protocol_engine/state/modules.py [hunk 7]
```diff
@@ -360,3 +412,54 @@ def find_attached_module(
                     return m
 
         raise errors.ModuleNotAttachedError(f"No available {model.value} found.")
+
+    @staticmethod
+    def calculate_magnet_hardware_height(
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/modules.py]: Known TODO, but marking with review for ease: write tests for this.

Also, we could make this a non-static method and take the `module_id` rather than the model. Internally, it can still use `self.get_model`, but (as commented above) it makes the implementation collaborator need to do less.

It also might make sense to make a `self.get_magnetic_module_model(module_id)` method that this method can then call to. This will make the tests easier I think

File: api/tests/opentrons/protocol_engine/state/test_module_view.py [hunk 8]
```diff
@@ -251,12 +254,64 @@ def test_thermocycler_dodging(
...
 
 
-def test_find_attached_module_rejects_missing() -> None:
+def test_find_loaded_hardware_module(
+    decoy: Decoy,
+    magdeck_v1_def: ModuleDefinition
+) -> None:
+    """It should return the matching hardware module."""
+
+    matching = decoy.mock(cls=AbstractModule)
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_module_view.py]: Highlighting this because ideally we wouldn't use `Decoy` here, but also all of our existing hardware module classes are prohibitively difficult to instantiate, and this is pretty benign mock usage. Could we add a `NOTE` comment to highlight that this is suboptimal but out of our control?

File: api/src/opentrons/protocol_engine/state/modules.py [hunk 9]
```diff
@@ -360,3 +412,54 @@ def find_attached_module(
...
+            RuntimeError: If modules of the given model are physically incapable
+                of reaching the requested height.
+        """
+        if magnetic_module_model not in [
+            ModuleModel.MAGNETIC_MODULE_V1,
+            ModuleModel.MAGNETIC_MODULE_V2,
+        ]:
+            raise errors.WrongModuleTypeError(
+                f"{magnetic_module_model} is not a Magnetic Module."
+            )
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/modules.py]: This logic could be in a `self.get_magnetic_module_model` method

File: api/src/opentrons/protocol_engine/state/modules.py [hunk 10]
```diff
@@ -360,3 +412,54 @@ def find_attached_module(
...
+        ]
+        hardware_units_above_home = home_to_base_offset + hardware_units_above_base
+        if not engage_height_is_in_range(
+            model=magnetic_module_model, height=hardware_units_above_home
+        ):
+            # TODO(mm, 2022-03-02): This error message probably will not match how
+            # the user specified the height. (Hardware units versus mm,
+            # home as origin versus labware base as origin.) This may be confusing
+            # depending on how it propagates up.
+            raise errors.EngageHeightOutOfRangeError(
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/modules.py]: I would add the module model, mm from base, and limit value as arguments to this exception so we can put them in a payload as data later. Also would allow us to move this message construction out of here

**Review by mcous — COMMENTED** [review]
Reviewed with @sanni-t and @y3rsh in pairing+review meeting. Mostly little comments around naming and organization while the remaining tests are written

File: api/src/opentrons/protocol_engine/execution/equipment.py (hunk 4 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/execution/equipment.py]: Omitting this for now, for simplicity/YAGNI.

File: api/tests/opentrons/protocol_engine/state/test_module_view.py [hunk 11]
```diff
@@ -204,6 +222,86 @@ def test_calculate_magnet_height(module_model: ModuleModel) -> None:
...
+            # https://github.com/Opentrons/opentrons/issues/9585
+            expected_result=25,
+            expected_exception_type=None,
+        ),
+        _CalculateMagnetHardwareHeightTestParams(
+            model=ModuleModel.MAGNETIC_MODULE_V2,
+            mm_from_base=10,
+            expected_result=12.5,
+            expected_exception_type=None,
+        ),
```
Comment (SyntaxColoring) [inline api/tests/opentrons/protocol_engine/state/test_module_view.py:254]: Add boundary conditions

File: api/tests/opentrons/protocol_engine/state/test_module_view.py [hunk 12]
```diff
@@ -251,12 +349,123 @@ def test_thermocycler_dodging(
     )
 
 
-def test_find_attached_module_rejects_missing() -> None:
+def test_find_loaded_hardware_module(
+    decoy: Decoy, magdeck_v1_def: ModuleDefinition
+) -> None:
+    """It should return the matching hardware module."""
+    matching = make_hardware_module(decoy=decoy, serial_number="serial-matching")
```
Comment (SyntaxColoring) [inline api/tests/opentrons/protocol_engine/state/test_module_view.py]: Find a way to use concrete classes here (e.g. `MagDeck`, `TempDeck`) instead of `AbstractModule`.

File: api/src/opentrons/protocol_engine/commands/magnetic_module/engage.py [hunk 13]
```diff
@@ -51,12 +59,57 @@ class EngageResult(BaseModel):
...
+        # Allow propagation of ModuleDoesNotExistError.
+        model = self._state_view.modules.get_model(module_id=magnetic_module_id)
+
+        # Allow propagation of WrongModuleTypeError and EngageHeightOutOfRangeError.
+        hardware_height = self._state_view.modules.calculate_magnet_hardware_height(
+            magnetic_module_model=model,
+            mm_from_base=mm_from_base,
         )
 
+        if not self._state_view.get_configs().use_virtual_modules:
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/magnetic_module/engage.py:104]: Test both branches.

File: api/src/opentrons/protocol_engine/state/modules.py (hunk 7 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/state/modules.py]: > Known TODO, but marking with review for ease: write tests for this.

👍 

> Also, we could make this a non-static method and take the module_id rather than the model. Internally, it can still use self.get_model, but (as commented above) it makes the implementation collaborator need to do less.

Discussed in call: not doing this for now because I softly preferred this logic in its own portable method, isolated from Protocol Engine notions. (Even though it lives within Protocol Engine.) This is a soft preference and I wouldn't mind if it ends up the way you suggested in another PR.

File: api/src/opentrons/protocol_engine/execution/equipment.py (hunk 3 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/execution/equipment.py]: See https://github.com/Opentrons/opentrons/pull/9599#discussion_r819667540

**Review by mcous — APPROVED** [review]
Code and tests LGTM! Based on talking with Josh during review session this morning, I think we're good to merge this and continue work on disengage
